File: apps/flutter/lib/di/backend.dart
/// The composition root's dependency graph: demo (in-memory) or real (HTTP + session).
library;

import 'dart:math';

import '../data/api.dart';
import '../data/auth.dart';
import '../data/cache.dart';
import '../demo/demo.dart';
import '../domain/repositories.dart';

/// [session] is null in demo mode (no auth); the app shows the login screen only when a real
/// [SessionManager] is present and has no token.
class Backend {
  final CachingEventRepository events;
  final QueueRepository queue;
  final ReservationRepository reservations;
  final OrderRepository orders;
  final IdempotencyKeyFactory keys;
  final SessionManager? session;
  Backend({
    required this.events,
    required this.queue,
    required this.reservations,
    required this.orders,
    required this.keys,
    this.session,
  });
}

/// In-memory data, no auth. Runs with no backend.
Backend demoBackend() => Backend(
      events: CachingEventRepository(DemoEventRepository()),
      queue: DemoQueueRepository(),
      reservations: DemoReservationRepository(),
      orders: DemoOrderRepository(),
      keys: DemoIdempotencyKeyFactory(),
    );

/// Real HTTP repositories against the gateway, with a session and refresh rotation.
Backend realBackend(ApiConfig config) {
  // Auth calls go through a session-less executor so refresh does not carry a stale token.
  final authExecutor = ApiExecutor(buildDio(config));
  final session = SessionManager(InMemoryTokenStore(), HttpAuthRepository(authExecutor));
  final executor = ApiExecutor(buildDio(config), session: session);
  return Backend(
    events: CachingEventRepository(HttpEventRepository(executor)),
    queue: HttpQueueRepository(executor),
    reservations: HttpReservationRepository(executor),
    orders: HttpOrderRepository(executor),
    keys: _RandomKeyFactory(),
    session: session,
  );
}

class _RandomKeyFactory implements IdempotencyKeyFactory {
  final Random _random = Random.secure();
  @override
  String newKey() {
    final bytes = List<int>.generate(16, (_) => _random.nextInt(256));
    return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }
}

File: apps/kmp/sharedUI/src/commonMain/kotlin/com/ticketinglabs/client/di/Backend.kt
package com.ticketinglabs.client.di

import com.ticketinglabs.client.data.ApiConfig
import com.ticketinglabs.client.data.auth.SessionManager
import com.ticketinglabs.client.data.auth.createSecureTokenStore
import com.ticketinglabs.client.data.cache.CachingEventRepository
import com.ticketinglabs.client.data.http.ApiExecutor
import com.ticketinglabs.client.data.http.HttpClientFactory
import com.ticketinglabs.client.data.repository.HttpAuthRepository
import com.ticketinglabs.client.data.repository.HttpEventRepository
import com.ticketinglabs.client.data.repository.HttpOrderRepository
import com.ticketinglabs.client.data.repository.HttpQueueRepository
import com.ticketinglabs.client.data.repository.HttpReservationRepository
import com.ticketinglabs.client.demo.DemoEventRepository
import com.ticketinglabs.client.demo.DemoIdempotencyKeyFactory
import com.ticketinglabs.client.demo.DemoOrderRepository
import com.ticketinglabs.client.demo.DemoQueueRepository
import com.ticketinglabs.client.demo.DemoReservationRepository
import com.ticketinglabs.client.domain.port.IdempotencyKeyFactory
import com.ticketinglabs.client.domain.port.OrderRepository
import com.ticketinglabs.client.domain.port.QueueRepository
import com.ticketinglabs.client.domain.port.ReservationRepository
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/**
 * The composition root's dependency graph. [session] is null in demo mode (no auth); the app
 * shows the login screen only when a real [SessionManager] is present and has no token.
 */
class Backend(
    val events: CachingEventRepository,
    val queue: QueueRepository,
    val reservations: ReservationRepository,
    val orders: OrderRepository,
    val keys: IdempotencyKeyFactory,
    val session: SessionManager?,
)

/** In-memory data, no auth. Runs with no backend. */
fun demoBackend(): Backend = Backend(
    events = CachingEventRepository(DemoEventRepository()),
    queue = DemoQueueRepository(),
    reservations = DemoReservationRepository(),
    orders = DemoOrderRepository(),
    keys = DemoIdempotencyKeyFactory(),
    session = null,
)

/** Real HTTP repositories against the gateway, with a session and refresh rotation. */
fun realBackend(config: ApiConfig): Backend {
    val json = HttpClientFactory.defaultJson()
    // Auth calls go through a session-less executor so refresh does not carry a stale token.
    val authExecutor = ApiExecutor(HttpClientFactory.createDefault(config), json)
    val session = SessionManager(createSecureTokenStore(), HttpAuthRepository(authExecutor, json))
    val executor = ApiExecutor(HttpClientFactory.createDefault(config), json, session = session)
    return Backend(
        events = CachingEventRepository(HttpEventRepository(executor, json)),
        queue = HttpQueueRepository(executor, json),
        reservations = HttpReservationRepository(executor, json),
        orders = HttpOrderRepository(executor, json),
        keys = UuidKeyFactory(),
        session = session,
    )
}

@OptIn(ExperimentalUuidApi::class)
private class UuidKeyFactory : IdempotencyKeyFactory {
    override fun newKey(): String = Uuid.random().toString()
}
